src/cart/crt_hg.c

```diff
@@ -1516,6 +1516,16 @@ crt_hg_reply_send(struct crt_rpc_priv *rpc_priv)
 		rc = crt_hgret_2_der(hg_ret);
 	}
 
+	/* Release input buffer */
+	if (rpc_priv->crp_release_input_early && !rpc_priv->crp_forward) {
```

Comment:
What will happen if we release RPC input buffer only with checking **"if (!rpc_priv->crp_forward)"**? that will allow us to release input buffer for most of RPCs (except corpc children) without changing the callers' logic (such as the RPC handlers for object/dtx in this patch)? Although the observed timeout RPCs in DAOS-15914 logs were for object update and DTX commit, it does not means that the trouble RPCs holding multiple recv buffer were all for object update and DTX commit, instead, they may be for others, we do not know. So if we can release the input buffer for more RPCs, that may be more helpful.

Reply:
@Nasf-Fan we did that in the initial patch (https://github.com/daos-stack/daos/pull/14793/files) and ended up with random memory corruptions, suggesting that something in daos keeps using input buffer still during crt_hg_reply_send(). 

See Saurabh updates with this error in DAOS-15914:
`aurora-daos-0318 ERROR 2024/07/25 00:38:05 daos_engine:0 07/25-00:38:05.15 aurora-daos-0318 DAOS[61709/14/30252] common EMRG src/common/mem.c:1383 set_offsets() Assertion '!OID_IS_NULL(root_oid)' failed: You must call pmemobj_root before umem_class_init
daos_engine: src/common/mem.c:1383: set_offsets: Assertion `0' failed.`

Reply:
OK, I am not sure which RPC handler (since there are too many RPC handlers for kinds of modules) may access the input buffer after early release, but I think that releasing the input buffer only for tgt_update and dtx_commit RPCs seems not enough. I would suggest to do that for all OBJ/DTX sponsored RPCs. If still hit above memory corruption, then it is not difficult to fix related RPC handler (in such two modules).